plan/relations.go

```diff
@@ -449,6 +541,40 @@ func (p *ProjectRel) ToProtoPlanRel() *proto.PlanRel {
 	}
 }
 
+func (p *ProjectRel) GetInputs() []Rel {
+	return []Rel{p.input}
+}
+
+func (p *ProjectRel) Copy(newInputs ...Rel) (Rel, error) {
+	if len(newInputs) != 1 {
+		return nil, substraitgo.ErrInvalidInputCount
+	}
+	proj := *p
+	proj.input = newInputs[0]
+	return &proj, nil
+}
+
+func (p *ProjectRel) CopyWithExpressionRewrite(rewriteFunc func(expr.Expression) (expr.Expression, error), newInputs ...Rel) (Rel, error) {
+	if len(newInputs) != 1 {
+		return nil, substraitgo.ErrInvalidInputCount
+	}
+	exprs := make([]expr.Expression, len(p.exprs))
+	for i, e := range p.exprs {
+		newExpr, err := rewriteFunc(e)
+		if err != nil {
+			return nil, err
+		}
+		exprs[i] = newExpr
+	}
+	if areExpressionsEqual(exprs, p.exprs) && areRelsEqual(newInputs, p.GetInputs()) {
+		return nil, nil
+	}
```

Comment:
If these are only checking that the underlying pointers the same you can just use `slices.Equal(exprs, p.exprs)` instead and avoid having to make your own functions for this. Same goes for the rels.